DBZ-7762 Document connector-specific custom converters (#5471)
debezium/debezium: documentation/modules/ROOT/pages/connectors/mysql.adoc

|===

[[debezium-mysql-connector-converters]]
== Custom converters

The {prodname} MySQL connector provides several `CustomConverter` implementations specific for MySQL out of the box.

> These custom converters provide alternative mappings for specific data types influenced by configuration.

Author: So what I am trying to convey here is that out of the box, the standard connector value converters have a set of default mappings from JDBC/column types to Debezium/Kafka logic types.  

However, users may not find these mappings ideal for their specific use case and may want to instead emit certain column types differently.  They supplement their connector configuration with these converter configurations to alter the default behavior based on the converters they enable.

To add a `CustomConverter` to the connector, follow the instructions in the link:../development/converters.adoc[Custom Converters documentation].

=== `TINYINT(1)` to Boolean

> MySQL reports `BOOLEAN` data types as `TINYINT(1)` during the connector's snapshot.

Author: Hi @roldanbob, so during the snapshot we gather column types from the JDBC driver, and during the streaming phase, this is gathered from the binlog events themselves and these two are not consistent.

> MySQL reports `BOOLEAN` data types as `TINYINT(1)` during the connector's snapshot.

Reviewer: Revised.

> This may not be ideal as it will create a difference in the change event schema with the streaming phase.

Reviewer: ```suggestion
After the connector transitions from the snapshot to the streaming phase, the change event schema that results from the default mapping can lead to inconsistent mappings for `BOOLEAN` columns. 
```

Reviewer: Revised.

> To make sure that MySQL emits `BOOLEAN` columns uniformly, the `TinyIntOneToBooleanConverter` can be used.

Reviewer: ```suggestion
To help ensure that MySQL emits `BOOLEAN` columns uniformly, you can apply the custom  `TinyIntOneToBooleanConverter`, as shown in the following configuration example.
```
In general, the lawyers cringe if the docs imply that anything is `sure` or `guaranteed`. But we can _help_ in that regard.

> .Example configuration

Reviewer: ```suggestion
.Example: `TinyIntOneToBooleanConverter` configuration 
```

[source]
----
converters=tinyint-one-to-boolean
converters.tinyint-one-to-boolean.type=io.debezium.connector.mysql.converters.TinyIntOneToBooleanConverter
converters.tinyint-one-to-boolean.selector=.*.MY_TABLE.DATA
converters.tinyint-one-to-boolean.length.checker=false
----

> In this example, the `selector` and `length.checker` properties are optional.

Reviewer: ```suggestion
In the preceding example, the `selector` and `length.checker` properties are optional.
```

> When specifying the `length.checker` as `false`, the converter will not explicitly confirm that the `TINYINT` data type conforms to a length of `1`.
> By default, the length is checked.

Reviewer: ```suggestion
By default, the converter checks that `TINYINT` data types conform to a length of `1`.
Set `length.checker` to `false` to prevent the converter from explicitly confirming the length.
```

> When specifying the `selector`, this designates which tables or columns will be converted based on the supplied regular expression.

Reviewer: ```suggestion
The `selector` designates the tables or columns to convert, based on the supplied regular expression.
```

> If the `selector` property is omitted, the converter will convert all `TINYINT` columns to logical `BOOL` field types, restricting this to only `TINYINT(1)` if the `length.checker` property is omitted or set to `true`.

Reviewer: ```suggestion
If you omit the `selector` property, the converter maps all `TINYINT` columns to logical `BOOL` field types.
To map `TINYINT` columns to `TINYINT(1)` when the `selector` option is not present , omit the `length.checker` property, or set its value to `true`.
```

=== JDBC sink data types

> When integrating the {prodname} JDBC sink connector with a {prodname} MySQL source connector, some column attributes are emitted differently between the snapshot and streaming phases of the connector.

Reviewer: ```suggestion
If you integrate the {prodname} JDBC sink connector with a {prodname} MySQL source connector, the MYSQL connector emits some column attributes differently during the snapshot and streaming phases.
```

For the JDBC sink connector to consistently consume changes from both the snapshot and streaming phase, it's imperative to include this converter as part of the MySQL source connector configuration.

> .Example configuration

Reviewer: ```suggestion
.Example: `JdbcSinkDataTypesConverter` configuration
```

[source]
----
converters=jdbc-sink
converters.jdbc-sink.type=io.debezium.connector.mysql.converters.JdbcSinkDataTypesConverter
converters.jdbc-sink.selector.boolean=.*.MY_TABLE.BOOL_COL
converters.jdbc-sink.selector.real=.*.MY_TABLE.REAL_COL
converters.jdbc-sink.selector.string=.*.MY_TABLE.STRING_COL
converters.jdbc-sink.treat.real.as.double=true
----

> In this example, all `selector.*` and `treat.real.as.double` configuration properties are optional.

Reviewer: ```suggestion
In the preceding example, the `selector.*` and `treat.real.as.double` configuration properties are optional.
```

> The `selector.*` properties allow specifying a comma-separated list of regular expressions to limit the table and columns that are converted.

Reviewer: ```suggestion
The `selector.*` properties specify comma-separated lists of regular expressions that specify which tables and columns that the converter applies to.
```

> If a specific selector property is omitted for a given field type, then all fields matching that rule will be converted.

Reviewer: ```suggestion
If you omit a specific selector property for a field type, the converter processes all fields that match that rule.
```

> Additionally, the converter will always treat `REAL` data types as `FLOAT64` logical values, unless `treat.real.as.double` is set to `false` causing the field to be emitted as `FLOAT32` instead.

Reviewer: ```suggestion
By default, the converter treats `REAL` data types as `FLOAT64` logical values.
To emit `REAL` types as `FLOAT32` types, set `treat.real.as.double` to `false`.
```

> The converter applies the following rules:

Reviewer: ```suggestion
By default, the converter applies the following rules:
```

* `BOOLEAN` data types are always emitted as `INT16` logical types, with `1` representing `true` and `0` representing `false`
* `REAL` data types are always emitted as `FLOAT64` logical types.

> * String-based columns always include the `__debezium.source.column.character_set` schema parameter containing the column's character set.

Reviewer: ```suggestion
* String-based columns always include the `__debezium.source.column.character_set` schema parameter that contains the column's character set.
```

|===

[[debezium-mysql-connector-converters]]
== Custom converters

By default, the {prodname} MySQL connector provides several `CustomConverter` implementations for MySQL data types.
These custom converters provide alternative mappings for specific data types based on the connector configuration.
To add a `CustomConverter` to the connector, follow the instructions in the link:../development/converters.adoc[Custom Converters documentation].

=== `TINYINT(1)` to Boolean

> By default, during a connector snapshot, {prodname} obtains column types from the JDBC driver, which assigns the `TINYINT(1)` type to BOOLEAN` columns.

Author: Well in this context we're specifically referring to MySQL so I think it makes sense to add it that way :+1:

To help ensure that MySQL emits `BOOLEAN` columns uniformly, you can apply the custom  `TinyIntOneToBooleanConverter`, as shown in the following configuration example.

.Example: `TinyIntOneToBooleanConverter` configuration
[source]
----
converters=tinyint-one-to-boolean
converters.tinyint-one-to-boolean.type=io.debezium.connector.mysql.converters.TinyIntOneToBooleanConverter
converters.tinyint-one-to-boolean.selector=.*.MY_TABLE.DATA
converters.tinyint-one-to-boolean.length.checker=false
----

In the preceding example, the `selector` and `length.checker` properties are optional.
By default, the converter checks that `TINYINT` data types conform to a length of `1`.
If `length.checker` to `false`, the converter does not explicitly confirm that the `TINYINT` data type conforms to a length of `1`.
The `selector` designates the tables or columns to convert, based on the supplied regular expression.
If you omit the `selector` property, the converter maps all `TINYINT` columns to logical `BOOL` field types.

> To map `TINYINT` columns to `TINYINT(1)` when the `selector` option is not present, omit the `length.checker` property, or set its value to `true`.

Reviewer: ```suggestion
If you do not configure a `selector` option, and you want to map `TINYINT` columns to `TINYINT(1)`, omit the `length.checker` property, or set its value to `true`.
```

To add a `CustomConverter` to the connector, follow the instructions in the link:../development/converters.adoc[Custom Converters documentation].

=== `TINYINT(1)` to Boolean

By default, during a connector snapshot, {prodname} obtains column types from the JDBC driver, which assigns the `TINYINT(1)` type to BOOLEAN` columns.
{prodname} then uses these JDBC column types to define the schema for the snapshot events.
After the connector transitions from the snapshot to the streaming phase, the change event schema that results from the default mapping can lead to inconsistent mappings for `BOOLEAN` columns.
To help ensure that MySQL emits `BOOLEAN` columns uniformly, you can apply the custom  `TinyIntOneToBooleanConverter`, as shown in the following configuration example.

.Example: `TinyIntOneToBooleanConverter` configuration
[source]
----
converters=tinyint-one-to-boolean
converters.tinyint-one-to-boolean.type=io.debezium.connector.mysql.converters.TinyIntOneToBooleanConverter
converters.tinyint-one-to-boolean.selector=.*.MY_TABLE.DATA
converters.tinyint-one-to-boolean.length.checker=false
----

In the preceding example, the `selector` and `length.checker` properties are optional.

> When specifying the `length.checker` as `false`, the converter will not explicitly confirm that the `TINYINT` data type conforms to a length of `1`.
> By default, the converter checks that `TINYINT` data types conform to a length of `1`.
> If `length.checker` to `false`, the converter does not explicitly confirm that the `TINYINT` data type conforms to a length of `1`.

Reviewer: ```suggestion
By default, the converter checks that `TINYINT` data types conform to a length of `1`.
If `length.checker` to `false`, the converter does not explicitly confirm that the `TINYINT` data type conforms to a length of `1`.
```

The `selector` designates the tables or columns to convert, based on the supplied regular expression.
If you omit the `selector` property, the converter maps all `TINYINT` columns to logical `BOOL` field types.

> To map `TINYINT` columns to `TINYINT(1)` when the `selector` option is not present , omit the `length.checker` property, or set its value to `true`.

Reviewer: ```suggestion
To map `TINYINT` columns to `TINYINT(1)` when the `selector` option is not present, omit the `length.checker` property, or set its value to `true`.
```

converters.jdbc-sink.type=io.debezium.connector.mysql.converters.JdbcSinkDataTypesConverter
converters.jdbc-sink.selector.boolean=.*.MY_TABLE.BOOL_COL
converters.jdbc-sink.selector.real=.*.MY_TABLE.REAL_COL
converters.jdbc-sink.selector.string=.*.MY_TABLE.STRING_COL
converters.jdbc-sink.treat.real.as.double=true
----

In the preceding example, the `selector.*` and `treat.real.as.double` configuration properties are optional.

The `selector.*` properties specify comma-separated lists of regular expressions that specify which tables and columns that the converter applies to.
By default, the converter applies the following rules apply to all Boolean, real, and string-based column data types, across all tables:

* `BOOLEAN` data types are always emitted as `INT16` logical types, with `1` representing `true` and `0` representing `false`
* `REAL` data types are always emitted as `FLOAT64` logical types.
* String-based columns always include the `__debezium.source.column.character_set` schema parameter that contains the column's character set.

For each data type, you can configure a selector rule to override the default scope and apply the selector to specific tables and columns only.

> For example, to set the scope of the Boolean converter, add the following rule to the connector configuration, as in the  preceding example: `converters.jdbc-sink.selector.boolean=.*.MY_TABLE.BOOL_COL`

Reviewer: ```suggestion
For example, to set the scope of the Boolean converter, add the following rule to the connector configuration, as in the preceding example: `converters.jdbc-sink.selector.boolean=.*.MY_TABLE.BOOL_COL`
```

=== JDBC sink data types

If you integrate the {prodname} JDBC sink connector with a {prodname} MySQL source connector, the MYSQL connector emits some column attributes differently during the snapshot and streaming phases.
For the JDBC sink connector to consistently consume changes from both the snapshot and streaming phase, you must include the `JdbcSinkDataTypesConverter` converter as part of the MySQL source connector configuration, as shown in the following example:

.Example: `JdbcSinkDataTypesConverter` configuration
[source]
----
converters=jdbc-sink
converters.jdbc-sink.type=io.debezium.connector.mysql.converters.JdbcSinkDataTypesConverter
converters.jdbc-sink.selector.boolean=.*.MY_TABLE.BOOL_COL
converters.jdbc-sink.selector.real=.*.MY_TABLE.REAL_COL
converters.jdbc-sink.selector.string=.*.MY_TABLE.STRING_COL
converters.jdbc-sink.treat.real.as.double=true
----

In the preceding example, the `selector.*` and `treat.real.as.double` configuration properties are optional.

The `selector.*` properties specify comma-separated lists of regular expressions that specify which tables and columns that the converter applies to.

> If you omit a specific selector property for a field type, the converter processes all fields that match that rule.

Reviewer: ```suggestion
If you do not specify a `selector` property for a field type, when {prodname} emits an event, the converter uses the rule that you specify in the configuration to map the data types of all columns of that type.
```
Taking another stab at this, bc, reading through L2254 again, I realize that it's not quite clear. I don't think that the original text, is any more clear: _"If a specific selector property is omitted for a given field type, then all fields matching that rule will be converted"_

So, if I don't set a `selector` for columns of a specific data type, the existing sentence states that all fields that match some `rule` are converted. 
I'd like to explicitly call out what `rule` we're referring to:

- The default rules for `BOOLEAN` and `REAL` data types that we mention in the paragraphs that follow. 
- The rule defined by the `treat.real.as.double` property.

Based on the current order in which we present the information, it's ambiguous as to whether `rule` might refer to the regex defined for the selector, -- which, of course makes `0` sense in the context of the case where no selector is set.  

I wonder whether we should start this section with a description of the main/default function of the converter (the information in L2255-2262). Or at least explicitly reference the "default converter rules" or the specific converter function that you define in the configuration (is this limited to `treat.real.as.double` or are there others?).

Author: So these `selectors` we we refer to have a default rule that accepts all combinations.  So, in this case, if the user does not provide a `selector.boolean` configuration option, by default, we match all boolean data types.  So the net effect is that the user can reduce the scope of the selector by providing their own rule if that's preferred.

Reviewer: ```suggestion
By default, the converter applies the following rules apply to all Boolean, real, and string-based column data types, across all tables:

* `BOOLEAN` data types are always emitted as `INT16` logical types, with `1` representing `true` and `0` representing `false`
* `REAL` data types are always emitted as `FLOAT64` logical types.
* String-based columns always include the `__debezium.source.column.character_set` schema parameter that contains the column's character set.

For each data type, you can configure a selector rule to override the default scope and apply the selector to specific tables and columns only.
For example, to set the scope of the Boolean converter, add the following rule to the connector configuration, as in the  preceding example: `converters.jdbc-sink.selector.boolean=.*.MY_TABLE.BOOL_COL`
```

OK. I think that we can do something like this ^^ 

I wanted to leave the existing conversation comments intact, so I didn't replace those, but maybe we can use this in place of L2254-2263?
